Openssh 8.1

julek-wolfssl opened the pull request:
These are changes necessary to get OpenSSH 8.1 to compile and pass tests with wolfSSL. After this code is merged I will open a PR with OpenSSH to try to get wolfSSL compatibility code in the master branch.

SparkiDev commented on src/ssl.c:
```diff
@@ -15569,6 +15571,10 @@ size_t wolfSSL_get_client_random(const WOLFSSL* ssl, unsigned char* out,
         return SSLEAY_VERSION_NUMBER;
     }
 
+    unsigned long OpenSSL_version_num(void)
```
Make this wolfSSL_OpenSSL_version_num.

SparkiDev commented on src/ssl.c:
```diff
@@ -25393,6 +25540,23 @@ unsigned long wolfSSL_ERR_peek_error(void)
     return wolfSSL_ERR_peek_error_line_data(NULL, NULL, NULL, NULL);
 }
 
+int wolfSSL_ERR_GET_LIB(unsigned long err)
```
OpenSSL uses a part of the number to indicate the library.
Can we do this too?

julek-wolfssl replied:
I think that it would make recognizing our errors difficult. In my opinion its easier to switch case it here and not worry about OpenSSL <-> wolfSSL error number conversions.

SparkiDev commented on src/ssl.c:
```diff
@@ -30480,6 +30649,30 @@ int wolfSSL_DSA_generate_parameters_ex(WOLFSSL_DSA* dsa, int bits,
     return ret;
 }
 
+WOLFSSL_DSA_SIG* wolfSSL_DSA_SIG_new(void)
```
Make sure this only compiles when !NO_DSA.

julek-wolfssl replied:
There is an `#ifndef NO_DSA` on line 30480.

SparkiDev commented on src/ssl.c:
```diff
@@ -31782,6 +32066,51 @@ int wolfSSL_EVP_PKEY_assign(WOLFSSL_EVP_PKEY *pkey, int type, void *key)
+#ifdef HAVE_ECC
+    if((pkey == NULL) || (key ==NULL))return WOLFSSL_FAILURE;
+    WOLFSSL_ENTER("wolfSSL_EVP_PKEY_set1_EC_KEY");
+    if (pkey->rsa != NULL && pkey->ownRsa == 1) {
```
Does this need a !NO_RSA check?
What other fields are there? dsa?

SparkiDev commented on src/ssl.c:
```diff
@@ -32203,6 +32532,56 @@ static int EncryptDerKey(byte *der, int *derSz, const EVP_CIPHER* cipher,
 #endif /* WOLFSSL_KEY_GEN || WOLFSSL_PEM_TO_DER */
 
+#ifndef NO_RSA
+static int wolfSSL_RSA_To_Der(WOLFSSL_RSA* rsa, byte** outBuf, int publicKey)
```
Use wolfSSL_i2d_RSAPrivateKey() and wolfSSL_i2d_RsaPublicKey().
I think those functions are the real ones and we don't want to have public key code and private key code being include regardless of usage.

julek-wolfssl replied:
Can you take a look at what I did with these functions? Done in commit https://github.com/wolfSSL/wolfssl/pull/2733/commits/f07c1f947fdc4480f6a5ae0ed2e5927773486f36

SparkiDev replied:
It's not going to matter anyway as the OpenSSL APIs pull everything in.

SparkiDev commented on src/ssl.c:
```diff
@@ -33361,6 +33718,78 @@ int wolfSSL_EC_KEY_set_public_key(WOLFSSL_EC_KEY *key,
+    if (wc_ecc_sign_hash(digest, digestSz, sig, sigSz, rng, (ecc_key*)key->internal) != MP_OKAY) {
+        ret = WOLFSSL_FAILURE;
+    }
+    if (initTmpRng) {
```
Conditional code when WOLFSSL_SMALL_STACK and check tmpRNG != NULL

SparkiDev commented on src/ssl.c:
```diff
@@ -36001,6 +36597,68 @@ void wolfSSL_RSA_set_flags(WOLFSSL_RSA *r, int flags)
+{
+    WOLFSSL_ENTER("wolfSSL_RSA_get_ex_data");
+    #ifdef HAVE_EX_DATA
+    if(rsa) {
```
formatting

julek-wolfssl replied:
Can you take a look again if its ok?

SparkiDev commented on src/ssl.c:
```diff
@@ -45875,14 +46577,68 @@ int wolfSSL_BN_set_bit(WOLFSSL_BIGNUM* bn, int n)
+        if (mp_set_bit(tmp, n) != MP_OKAY) {
+            goto cleanup;
+        }
+        if (mp_sub((mp_int*)bn->internal, tmp, res) != MP_OKAY) {
```
Why use res?

julek-wolfssl replied:
Can parameters `b == c` in mp_sub? If not then it is necessary to have another variable to store the results.

julek-wolfssl replied:
Can `b == c` in mp_sub? If these parameters cannot be the same pointer then it is necessary to use `res`.

SparkiDev replied:
I would assume that a == c or b == c will work!
Test it out.

julek-wolfssl replied:
Looks like `a == c` works.

SparkiDev commented on wolfssl/wolfcrypt/asn.h:
```diff
@@ -220,7 +220,9 @@ enum
     NID_emailAddress = 0x30,           /* emailAddress */
     NID_id_on_dnsSRV = 82,             /* 1.3.6.1.5.5.7.8.7 */
-    NID_ms_upn = 265                   /* 1.3.6.1.4.1.311.20.2.3 */
+    NID_ms_upn = 265,                   /* 1.3.6.1.4.1.311.20.2.3 */
```
formatting - comment doesn't line up

SparkiDev commented on wolfssl/wolfcrypt/asn.h:
```diff
@@ -220,7 +220,9 @@ enum
-    NID_ms_upn = 265                   /* 1.3.6.1.4.1.311.20.2.3 */
+    NID_ms_upn = 265,                   /* 1.3.6.1.4.1.311.20.2.3 */
+
+    NID_X9_62_prime_field = 406
```
Add comment

SparkiDev commented:
Lots of work in here! Well done.
A few comments and testing failures to fix.

julek-wolfssl commented:
Thanks! I'm currently working on getting the test builds to pass and then I'll address your comments ASAP.

toddouska commented on src/wolfio.c:
```diff
@@ -770,7 +770,7 @@ int wolfIO_TcpConnect(SOCKET_T* sockfd, const char* ip, word16 port, int to_sec)
     SOCKADDR_S addr;
     int sockaddr_len = sizeof(SOCKADDR_IN);
-#ifdef HAVE_GETADDRINFO
+#ifndef WOLF_C99
```
This is not portable.  Before, autoconf detected whether getaddrinfo was available.  Now, an embedded user, who probably won't turn on WOLF_C99  is going to have code compiled that assumes getaddrinfo is available.

toddouska commented on wolfcrypt/src/asn.c:
```diff
@@ -69,6 +69,10 @@ ASN Options:
 #include <wolfssl/wolfcrypt/logging.h>
 
+#ifdef OPENSSL_EXTRA
+#include <wolfssl/openssl/ssl.h>
```
It's a mistake to bring in ssl headers down into our crypto layer.  Especially as we're trying to completely separate them.

julek-wolfssl commented:
In addition to your comments @toddouska I moved STR_SIZEOF to wolfssl/wolfcrypt/types.h and renamed it to XSTR_SIZEOF (following XMALLOC naming convention) to address possible collision concerns brought up by @dgarske.

JacobBarthelmeh commented on wolfssl/ssl.h:
```diff
@@ -69,7 +70,6 @@
         #include <openssl/err.h>
         #include <openssl/ec.h>
         #include <openssl/hmac.h>
-        #include <openssl/bn.h>
```
moving this and constantly include wolfssl/openssl/bn.h effects the opensslcoexist build
